README.md
# kafka-offset-exporter

----

## Description

This is a Prometheus exporter for topic and consumer group offsets in Kafka.
Your Kafka cluster must be on version 0.10.0.0 or above for this to work.

----

## Changelog

Please refer to the `CHANGELOG` file

## Migration to Go Mod

As the origial version of this project is not maintained anymore, I decided to fork it and move it to `go mod`.
For a user point of view, there's nothing to do to use this project as usual.

## Build

Simply clone the repo and `go build`, or use `go get` :

```bash
go get github.com/prune998/kafka-offset-exporter
```

### Docker

There is a basic `Dockerfile` that will build and package the application. The final image is based on Alpine.

To build it yourself, just clone the repository and build it :

```bash
git clone https://github.com/prune998/kafka-offset-exporter
cd kafka-offset-exporter
docker build -t kafka-offset-exporter:latest .
docker run -ti --rm prune/kafka-offset-exporter:v1.0.5
```

There is a maintained image at `prune/kafka-offset-exporter:v1.0.5` :

```bash
docker pull prune/kafka-offset-exporter:v1.0.5
docker run -ti --rm prune/kafka-offset-exporter:v1.0.5
```

In both case, add your arguments on the commandline (see *Usage*):

```bash
docker run -ti --rm prune/kafka-offset-exporter:v1.0.5 -brokers localhost:9092
```

## Usage

The only required parameter is a set of brokers to bootstrap into the cluster.

Parameters can be set on the commandline (see below) or defined as *Environment Variables*, or both at the same time :

```bash
LEVEL=debug ./kafka-offset-exporter -brokers=localhost:9092
```

This is usefull when you're deploying in Kubernetes :

```yaml
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: kafka-offset-exporter
  labels:
    app: "kafka-offset-exporter"
    release: "v1.0.5"
spec:
  replicas: 1
  template:
    metadata:
      labels:
        app: "kafka-offset-exporter"
        release: "v1.0.5"
    spec:
      containers:
      - name: kafka-offset-exporter
        image: "prune/kafka-offset-exporter:v1.0.5"
        imagePullPolicy: "Always"
        args:
        - "-brokers"
        - "localhost:9092"
        env:
        - name: LEVEL
          value: "debug"
```

By default, the oldest and newest offsets of all topics are retrieved and
reported.  You can also enable offset reporting for any consumer group but note
that due to the way Sarama works, this requires querying for offsets for _all_
partitions for each consumer group which can take a long time. It is recommended
to filter both topics and consumer groups to just the ones you care about.

```bash
$ ./kafka-offset-exporter -help
Usage of ./kafka-offset-exporter:
  -brokers string
        Kafka brokers to connect to, comma-separated
  -fetchMax duration
        Max time before requesting updates from broker (default 40s)
  -fetchMin duration
        Min time before requesting updates from broker (default 15s)
  -groups string
        Also fetch offsets for consumer groups matching this regex (default none)
  -level string
        Logger level (default "warn")
  -endpoint string
        Path to export metrics on (default "/metrics")
  -port int
        Port to export metrics on (default 9000)
  -refresh duration
        Time between refreshing cluster metadata (default 1m0s)
  -topics string
        Only fetch offsets for topics matching this regex (default all)
```

### Scraping with Prometheus-Operator

Prometheus Operator deployed in Kubernetes use a `ServiceMonitor` Custom Resource to scrape Services (you can't scrape pods directly).

#### Service

Create the service in the same `Namespace` as the `kafka-offset-exporter` using `kubectl -n your_namespace apply -f kafka-offset-exporter-svc.yml`

```yaml
apiVersion: v1
kind: Service
metadata:
  labels:
    app: kafka-offset-exporter
    release: kafka
  name: kafka-offset-exporter
spec:
  ports:
  - name: http
    port: 9000
    protocol: TCP
    targetPort: 9000
  selector:
    app: kafka-offset-exporter
  sessionAffinity: None
  type: ClusterIP
```

#### ServiceMonitor

Create the serviceMonitor in the same `Namespace` as `Prometheus` (in the `monitoring` namespace) using `kubectl -n monitoring -f kafka-offset-exporter-svcmon.yml`

```yaml
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  labels:
    app: kafka-offset-exporter
    k8s-app: "true"
  name: kafka-offset-exporter
spec:
  endpoints:
  - interval: 5s
    port: http
  namespaceSelector:
    any: true
  selector:
    matchLabels:
      app: kafka-offset-exporter
```

## The Original of kafka-offset-exporter version no longer maintained

The original version of this project is located at <https://github.com/echojc/kafka-offset-exporter>
It is not maintained anymore, while this fork is. Please submit PRs here.

From the orignial author Jonathan Chow <https://github.com/echojc> :

```text
My employer used to rely heavily on Kafka and so I could dogfood and iterate on
this project regularly. Unfortunately, this is no longer the case and I don't
have the resources to maintain and/or develop this anymore.

I don't feel comfortable recommending people to use outdated and unmaintained
software, so please consider using an established fork, forking this yourself,
or creating a new-and-improved exporter as an alternative.

Thanks all!
```

----


CHANGELOG
# 20231027 - V1.0.5 - prune
updated Go version 1.21 and updated dependencies (Sarama -> IBM)

# 20190806 - V1.0.4 - prune 
the `-path` commandline option was set by the `PATH` env variable when not set, which could lead to improbable URL...
the `-path` (or PATH) option is now deprecated in favour of `-endpoint` (ENDPOINT env var)

# 20190806 - V1.0.3 - prune 
Set the default Path (Endpoint) to be `/metrics` to conform to default Prometheus URLS

# 20190806 - V1.0.2 - prune 

Ensure the exporter close the connexion to the broker when we can't retreive data. The test using the broker.Connected() just check if an Address exist for this broker, not it the wire is still working.
There are places of improvements for it to work cleaner... but who's got time ? 
I also noted that the global workflow is biased : when the exporter starts, it fetches the list of Brokers, then start scraping each of them. This scenario does not take care of addition or removal of other Brokers...


kafka.go
package main

import (
	"errors"
	"math/rand"
	"regexp"
	"strconv"
	"sync"
	"time"

	"github.com/IBM/sarama"
	"github.com/prometheus/client_golang/prometheus"
	log "github.com/sirupsen/logrus"
)

type scrapeConfig struct {
	FetchMinInterval        time.Duration
	FetchMaxInterval        time.Duration
	MetadataRefreshInterval time.Duration
	TopicsFilter            *regexp.Regexp
	GroupsFilter            *regexp.Regexp
}

func mustNewScrapeConfig(refresh time.Duration, fetchMin time.Duration, fetchMax time.Duration, topics string, groups string) scrapeConfig {
	topicsFilter, err := regexp.Compile(topics)
	if err != nil {
		log.Fatal(err)
	}

	// empty group should match nothing instead of everything
	if groups == "" {
		groups = ".^"
	}
	groupsFilter, err := regexp.Compile(groups)
	if err != nil {
		log.Fatal(err)
	}

	return scrapeConfig{
		FetchMinInterval:        fetchMin,
		FetchMaxInterval:        fetchMax,
		MetadataRefreshInterval: refresh,
		TopicsFilter:            topicsFilter,
		GroupsFilter:            groupsFilter,
	}
}

func startKafkaScraper(wg *sync.WaitGroup, shutdown chan struct{}, kafka sarama.Client, cfg scrapeConfig) {
	wg.Add(1)
	go refreshMetadataPeriodically(wg, shutdown, kafka, cfg)
	for _, broker := range kafka.Brokers() {
		wg.Add(1)
		go manageBroker(wg, shutdown, broker, kafka, cfg)
	}
}

func refreshMetadataPeriodically(wg *sync.WaitGroup, shutdown chan struct{}, kafka sarama.Client, cfg scrapeConfig) {
	defer wg.Done()

	log.WithField("interval", cfg.MetadataRefreshInterval.String()).
		Info("Starting metadata refresh thread")

	wait := time.After(0)
	for {
		select {
		case <-wait:
			log.Debug("Refreshing cluster metadata")
			if err := kafka.RefreshMetadata(); err != nil {
				log.WithField("error", err).Warn("Failed to update cluster metadata")
			}
		case <-shutdown:
			log.Info("Shutting down metadata refresh thread")
			return
		}
		wait = time.After(cfg.MetadataRefreshInterval)
	}
}

func manageBroker(wg *sync.WaitGroup, shutdown chan struct{}, broker *sarama.Broker, kafka sarama.Client, cfg scrapeConfig) {
	defer wg.Done()

	log.WithField("broker", broker.Addr()).
		WithField("interval.min", cfg.FetchMinInterval.String()).
		WithField("interval.max", cfg.FetchMaxInterval.String()).
		Info("Starting handler for broker")

	wait := time.After(0)
	for {
		select {
		case <-wait:
			log.WithField("broker", broker.Addr()).Debug("Updating metrics")

			// ensure broker is connected
			if err := connect(broker); err != nil {
				log.WithField("broker", broker.Addr()).
					WithField("error", err).
					Error("Failed to connect to broker")
				break
			}

			// fetch groups coordinated by this broker
			var groups []string
			groupsResponse, err := broker.ListGroups(&sarama.ListGroupsRequest{})
			if err != nil {
				log.WithField("broker", broker.Addr()).
					WithField("error", err).
					Error("Failed to retrieve consumer groups")
				broker.Close()
				break
			} else if groupsResponse.Err != sarama.ErrNoError {
				log.WithField("broker", broker.Addr()).
					WithField("error", groupsResponse.Err).
					Error("Failed to retrieve consumer groups")
				broker.Close()
				break
			} else {
				for group := range groupsResponse.Groups {
					if !cfg.GroupsFilter.MatchString(group) {
						log.WithField("group", group).Info("not found group")
						continue
					}

					groupCoordinator, err := kafka.Coordinator(group)
					if err != nil {
						log.WithField("broker", broker.Addr()).
							WithField("group", group).
							WithField("error", err).
							Warn("Failed to identify broker for consumer group")
						continue
					}

					if broker == groupCoordinator {
						groups = append(groups, group)
					}
				}
			}
			if len(groups) == 0 {
				log.WithField("broker", broker.Addr()).Debug("No consumer groups to fetch offsets for")
			}

			// build requests
			partitionCount := 0
			oldestRequest := sarama.OffsetRequest{}
			newestRequest := sarama.OffsetRequest{}
			var groupRequests []sarama.OffsetFetchRequest
			for _, group := range groups {
				groupRequests = append(groupRequests, sarama.OffsetFetchRequest{ConsumerGroup: group, Version: 1})
			}

			// fetch partitions led by this broker, and add all partitions to group requests
			topics, err := kafka.Topics()
			if err != nil {
				log.WithField("broker", broker.Addr()).
					WithField("error", err).
					Error("Failed to get topics")
				break
			}

			for _, topic := range topics {
				if !cfg.TopicsFilter.MatchString(topic) {
					continue
				}

				partitions, err := kafka.Partitions(topic)
				if err != nil {
					log.WithField("broker", broker.Addr()).
						WithField("topic", topic).
						WithField("error", err).
						Warn("Failed to get partitions for topic")
					continue
				}

				for _, partition := range partitions {
					// all partitions need to be added to group requests
					for i := range groupRequests {
						groupRequests[i].AddPartition(topic, partition)
					}

					partitionLeader, err := kafka.Leader(topic, partition)
					if err != nil {
						log.WithField("broker", broker.Addr()).
							WithField("topic", topic).
							WithField("partition", partition).
							WithField("error", err).
							Warn("Failed to identify broker for partition")
						continue
					}

					if broker == partitionLeader {
						oldestRequest.AddBlock(topic, partition, sarama.OffsetOldest, 1)
						newestRequest.AddBlock(topic, partition, sarama.OffsetNewest, 1)
						partitionCount++
					}
				}
			}

			if partitionCount == 0 {
				log.WithField("broker", broker.Addr()).Debug("No partitions for broker to fetch")
			}

			log.WithField("broker", broker.Addr()).
				WithField("partition.count", partitionCount).
				WithField("group.count", len(groupRequests)).
				Debug("Sending requests")

			requestWG := &sync.WaitGroup{}
			requestWG.Add(2 + len(groupRequests))
			go func() {
				defer requestWG.Done()
				handleTopicOffsetRequest(broker, &oldestRequest, "oldest", metricOffsetOldest)
			}()
			go func() {
				defer requestWG.Done()
				handleTopicOffsetRequest(broker, &newestRequest, "newest", metricOffsetNewest)
			}()
			for i := range groupRequests {
				go func(request *sarama.OffsetFetchRequest) {
					defer requestWG.Done()
					handleGroupOffsetRequest(broker, request, metricOffsetConsumer)
				}(&groupRequests[i])
			}
			requestWG.Wait()

		case <-shutdown:
			log.WithField("broker", broker.Addr()).Info("Shutting down handler for broker")
			return
		}

		min := int64(cfg.FetchMinInterval)
		max := int64(cfg.FetchMaxInterval)
		duration := time.Duration(rand.Int63n(max-min) + min)

		wait = time.After(duration)
		log.WithField("broker", broker.Addr()).
			WithField("interval.rand", duration.String()).
			Debug("Updated metrics and waiting for next run")
	}
}

func connect(broker *sarama.Broker) error {
	if ok, _ := broker.Connected(); ok {
		return nil
	}

	cfg := sarama.NewConfig()
	cfg.Version = sarama.V1_0_0_0
	if err := broker.Open(cfg); err != nil {
		return err
	}

	if connected, err := broker.Connected(); err != nil {
		return err
	} else if !connected {
		return errors.New("Unknown failure")
	}

	return nil
}

func handleGroupOffsetRequest(broker *sarama.Broker, request *sarama.OffsetFetchRequest, metric *prometheus.GaugeVec) {
	response, err := broker.FetchOffset(request)
	if err != nil {
		log.WithField("broker", broker.Addr()).
			WithField("group", request.ConsumerGroup).
			WithField("error", err).
			Error("Failed to request group offsets")
		return
	}

	for topic, partitions := range response.Blocks {
		for partition, block := range partitions {
			if block == nil {
				log.WithField("broker", broker.Addr()).
					WithField("group", request.ConsumerGroup).
					WithField("topic", topic).
					WithField("partition", partition).
					Warn("Failed to get data for group")
				continue
			} else if block.Err != sarama.ErrNoError {
				log.WithField("broker", broker.Addr()).
					WithField("group", request.ConsumerGroup).
					WithField("topic", topic).
					WithField("partition", partition).
					WithField("error", block.Err).
					Warn("Failed getting data for group")
				continue
			} else if block.Offset < 0 {
				continue
			}

			metric.With(prometheus.Labels{
				"topic":     topic,
				"partition": strconv.Itoa(int(partition)),
				"group":     request.ConsumerGroup,
			}).Set(float64(block.Offset))
		}
	}
}

func handleTopicOffsetRequest(broker *sarama.Broker, request *sarama.OffsetRequest, offsetType string, metric *prometheus.GaugeVec) {
	response, err := broker.GetAvailableOffsets(request)
	if err != nil {
		log.WithField("broker", broker.Addr()).
			WithField("offset.type", offsetType).
			WithField("error", err).
			Error("Failed to request topic offsets")
		return
	}

	for topic, partitions := range response.Blocks {
		for partition, block := range partitions {
			if block == nil {
				log.WithField("broker", broker.Addr()).
					WithField("topic", topic).
					WithField("partition", partition).
					Warn("Failed to get data for partition")
				continue
			} else if block.Err != sarama.ErrNoError {
				log.WithField("broker", broker.Addr()).
					WithField("topic", topic).
					WithField("partition", partition).
					WithField("error", block.Err).
					Warn("Failed getting offsets for partition")
				continue
			} else if len(block.Offsets) != 1 {
				log.WithField("broker", broker.Addr()).
					WithField("topic", topic).
					WithField("partition", partition).
					Warn("Got unexpected offset data for partition")
				continue
			}

			metric.With(prometheus.Labels{
				"topic":     topic,
				"partition": strconv.Itoa(int(partition)),
			}).Set(float64(block.Offsets[0]))
		}
	}
}


metrics.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"sync"
	"time"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promhttp"
	log "github.com/sirupsen/logrus"
)

var (
	metricOffsetOldest = prometheus.NewGaugeVec(
		prometheus.GaugeOpts{
			Name: "kafka_offset_oldest",
			Help: "Oldest offset for a partition",
		},
		[]string{
			"topic",
			"partition",
		},
	)
	metricOffsetNewest = prometheus.NewGaugeVec(
		prometheus.GaugeOpts{
			Name: "kafka_offset_newest",
			Help: "Newest offset for a partition",
		},
		[]string{
			"topic",
			"partition",
		},
	)
	metricOffsetConsumer = prometheus.NewGaugeVec(
		prometheus.GaugeOpts{
			Name: "kafka_offset_consumer",
			Help: "Current offset for a consumer group",
		},
		[]string{
			"topic",
			"partition",
			"group",
		},
	)
)

func init() {
	prometheus.MustRegister(metricOffsetOldest)
	prometheus.MustRegister(metricOffsetNewest)
	prometheus.MustRegister(metricOffsetConsumer)
}

func startMetricsServer(wg *sync.WaitGroup, shutdown chan struct{}, cfg serverConfig) {
	wg.Add(1)
	go func() {
		defer wg.Done()

		mux := http.NewServeMux()

		// healthz basic
		mux.HandleFunc("/healthz", func(w http.ResponseWriter, r *http.Request) {
			m := map[string]interface{}{"version": version, "status": "OK"}

			b, err := json.Marshal(m)
			if err != nil {
				http.Error(w, "cant marshal healthz json", http.StatusInternalServerError)
				return
			}

			w.Write(b)
		})

		mux.Handle(cfg.path, promhttp.Handler())
		srv := &http.Server{
			Addr:    fmt.Sprintf(":%d", cfg.port),
			Handler: mux,
		}
		go func() {
			log.WithField("port", cfg.port).
				WithField("path", cfg.path).
				Info("Starting metrics HTTP server")
			srv.ListenAndServe()
		}()

		<-shutdown
		log.Info("Shutting down metrics HTTP server")
		ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
		defer cancel()
		srv.Shutdown(ctx)
	}()
}


main.go
package main

import (
	"os"
	"os/signal"
	"strings"
	"sync"
	"syscall"
	"time"

	"github.com/IBM/sarama"
	"github.com/namsral/flag"
	log "github.com/sirupsen/logrus"
)

var (
	version = "no version set"
)

type serverConfig struct {
	port int
	path string
}

func main() {
	brokerString := flag.String("brokers", "", "Kafka brokers to connect to, comma-separated")
	topics := flag.String("topics", "", "Only fetch offsets for topics matching this regex (default all)")
	groups := flag.String("groups", "", "Also fetch offsets for consumer groups matching this regex (default none)")
	port := flag.Int("port", 9000, "Port to export metrics on")
	path := flag.String("endpoint", "/metrics", "Path to export metrics on")
	refresh := flag.Duration("refresh", 1*time.Minute, "Time between refreshing cluster metadata")
	fetchMin := flag.Duration("fetchMin", 15*time.Second, "Min time before requesting updates from broker")
	fetchMax := flag.Duration("fetchMax", 40*time.Second, "Max time before requesting updates from broker")
	level := flag.String("level", log.WarnLevel.String(), "the log level to display (debug,info,error,warning)")
	flag.Parse()

	mustSetupLogger(*level)
	serverConfig := mustNewServerConfig(*port, *path)
	scrapeConfig := mustNewScrapeConfig(*refresh, *fetchMin, *fetchMax, *topics, *groups)

	kafka := mustNewKafka(*brokerString)
	defer kafka.Close()

	enforceGracefulShutdown(func(wg *sync.WaitGroup, shutdown chan struct{}) {
		startKafkaScraper(wg, shutdown, kafka, scrapeConfig)
		startMetricsServer(wg, shutdown, serverConfig)
	})
}

func mustNewServerConfig(port int, path string) serverConfig {

	return serverConfig{
		port: port,
		path: path,
	}
}

func enforceGracefulShutdown(f func(wg *sync.WaitGroup, shutdown chan struct{})) {
	wg := &sync.WaitGroup{}
	shutdown := make(chan struct{})
	signals := make(chan os.Signal)
	signal.Notify(signals, syscall.SIGTERM, syscall.SIGINT, syscall.SIGHUP)
	go func() {
		<-signals
		close(shutdown)
	}()

	log.Info("Graceful shutdown enabled")
	f(wg, shutdown)

	<-shutdown
	wg.Wait()
}

func mustNewKafka(brokerString string) sarama.Client {
	brokers := strings.Split(brokerString, ",")
	for i := range brokers {
		brokers[i] = strings.TrimSpace(brokers[i])
		if !strings.ContainsRune(brokers[i], ':') {
			brokers[i] += ":9092"
		}
	}
	log.WithField("brokers.bootstrap", brokers).Info("connecting to cluster with bootstrap hosts")

	cfg := sarama.NewConfig()
	cfg.Version = sarama.V1_0_0_0
	cfg.ClientID = "kafka-offset-exporter"

	client, err := sarama.NewClient(brokers, cfg)
	if err != nil {
		log.Fatal(err)
	}

	var addrs []string
	for _, b := range client.Brokers() {
		addrs = append(addrs, b.Addr())
	}
	log.WithField("brokers", addrs).Info("connected to cluster")

	return client
}

func mustSetupLogger(level string) {
	logLevel, err := log.ParseLevel(level)
	if err != nil {
		logLevel = log.WarnLevel
	}

	log.SetLevel(logLevel)
	log.SetFormatter(&log.JSONFormatter{})
}
